edb/edgeql/desugar_group.py

```diff
@@ -147,7 +147,17 @@ def rewrite(el: qlast.GroupingElement) -> qlast.GroupingElement:
     for alias, path in alias_map.values():
         using.append(qlast.AliasedExpr(alias=alias, expr=path))
 
-    actual_keys = collect_grouping_atoms(by)
+    keys: Sequence[Tuple[str, Optional[qlast.Span]]] = (
+        collect_grouping_atoms(by)
+    )
+    actual_keys: ordered.OrderedSet[str] = ordered.OrderedSet()
+    for key, key_span in keys:
+        if key in actual_keys:
+            raise errors.QueryError(
+                f"BY clause cannot refer to two properties with the same name",
+                span=key_span,
+            )
+        actual_keys.add(key)
```

Comment:
I think this is actually somewhat overzealous, now that I think about it.
```
_localdev:scope> group Card { name, cost, element } by {.cost, (.cost, .element)};
edgedb error: QueryError: BY clause cannot refer to two properties with the same name
_localdev:scope> 
```
This ought to work, certainly. (Also, the span isn't getting reported.)

We *could* try to do something more careful, and consider whether something is being used in a compound key, but I don't think it's worth it. I now think it's fine to allow stuff like `{.foo, .foo}` as long as we disallow `{.foo, @foo}`.

(The former case is a little ambiguous, but we *do* at least report the correct grouping and the correct key value, even if we can't distinguish the two identical versions. The latter will always omit key information.)